src/operator/special_functions-inl.h

```diff
@@ -0,0 +1,248 @@
+  MSHADOW_XINLINE static DType polevl(DType x, const DType coef[], int N) {
+    DType ans;
+    DType const *p;
+    int i;
+
+    p = coef;
+    ans = *p++;
+
+    i = N;
+    do {
+      ans = ans * x  +  *p++;
+    } while ( --i );
+
+    return( ans );
+  }
+
+
+  /*
+   * Helper function for psi that handles double/float specific differences
+   * in the algorithm.
+   */
+  template<typename DType>
+  MSHADOW_XINLINE static DType psi_helper(DType s);
+
+  /*
+   *
+   *	Psi (digamma) function
+   *
+   *
+   * SYNOPSIS:
+   *
+   * float x, y, psif();
+   *
+   * y = psif( x );
+   *
+   *
+   * DESCRIPTION:
+   *
+   *              d      -
+   *   psi(x)  =  -- ln | (x)
+   *              dx
+   *
+   * is the logarithmic derivative of the gamma function.
+   * For integer x,
+   *                   n-1
+   *                    -
+   * psi(n) = -EUL  +   >  1/k.
+   *                    -
+   *                   k=1
+   *
+   * This formula is used for 0 < n <= 10.  If x is negative, it
+   * is transformed to a positive argument by the reflection
+   * formula  psi(1-x) = psi(x) + pi cot(pi x).
+   * For general positive x, the argument is made greater than 10
+   * using the recurrence  psi(x+1) = psi(x) + 1/x.
+   * Then the following asymptotic expansion is applied:
+   *
+   *                           inf.   B
+   *                            -      2k
+   * psi(x) = log(x) - 1/2x -   >   -------
+   *                            -        2k
+   *                           k=1   2k x
+   *
+   * where the B2k are Bernoulli numbers.
+   *
+   * ACCURACY:
+   *    Absolute error,  relative when |psi| > 1 :
+   * arithmetic   domain     # trials      peak         rms
+   *    IEEE      -33,0        30000      8.2e-7      1.2e-7
+   *    IEEE      0,33        100000      7.3e-7      7.7e-8
+   *
+   * ERROR MESSAGES:
+   *     message         condition      value returned
+   * psi singularity    x integer <=0      MAXNUMF
+   */
+  template<typename DType>
+  MSHADOW_XINLINE static DType psi(DType x) {
+    DType p, q, nz, s, w, y;
+    int i, n, negative;
+
+    DType EUL(0.57721566490153286061);
+    DType PI(3.14159265358979323846);
+
+    negative = 0;
+    nz = 0.0;
+
+    if ( x <= 0.0 ) {
+      negative = 1;
+      q = x;
+      p = std::floor(q);
+      if ( p == q ) {
+        return helper_numeric_limits<double>::max();
+      }
+      /* Remove the zeros of tan(PI x)
+       * by subtracting the nearest integer from x
+       */
+      nz = q - p;
+      if ( nz != 0.5 ) {
+        if ( nz > 0.5 ) {
+          p += 1.0;
+          nz = q - p;
+        }
+        nz = PI/std::tan(PI*nz);
+      } else {
+        nz = 0.0;
+      }
+      x = 1.0 - x;
+    }
+
+    /* check for positive integer up to 10 */
+    if ( (x <= 10.0) && (x == std::floor(x)) ) {
+      y = 0.0;
+      n = x;
+      for ( i = 1; i < n; i++ ) {
+        w = i;
+        y += 1.0/w;
+      }
+      y -= EUL;
+      goto done;
```

Comment:
I think this should be fine. In my mind, it is better to stay with the code they provided so as to easily change it when they do. If you think it could be re-written in a more maintainable way, I'd encourage it.